pakyow-core/lib/pakyow/core/router.rb

```diff
@@ -1,174 +1,883 @@
-require 'singleton'
+require "pakyow/support/aargv"
+require "pakyow/support/array"
+
+require "pakyow/core/routing/hook_merger"
 
 module Pakyow
-  # A singleton that manages route sets.
+  # Executes code for particular requests. For example:
+  #
+  #   Pakyow::App.router do
+  #     get "/" do
+  #       # called for GET / requests
+  #     end
+  #   end
+  #
+  # A +Class+ is created dynamically for each defined router. When matched, a route
+  # is called in context of its router. This means that any method defined in a
+  # router is available to be called from within a route. For example:
+  #
+  #   Pakyow::App.router do
+  #     def foo
+  #     end
+  #
+  #     get :foo, "/foo" do
+  #       foo
+  #     end
+  #   end
+  #
+  # Including modules works as expected:
+  #
+  #   module AuthHelpers
+  #     def current_user
+  #     end
+  #   end
+  #
+  #   Pakyow::App.router do
+  #     include AuthHelpers
+  #
+  #     get :foo, "/foo" do
+  #       current_user
+  #     end
+  #   end
+  #
+  # See {App.router} for more details on defining routers.
+  #
+  # = Supported HTTP methods
+  #
+  # - +GET+
+  # - +POST+
+  # - +PUT+
+  # - +PATCH+
+  # - +DELETE+
+  #
+  # See {get}, {post}, {put}, {patch}, and {delete}.
+  #
+  # +HEAD+ requests are handled automatically via {Rack::Head}.
+  #
+  # = Building paths for named routes
+  #
+  # Path building is supported via {Controller#path} and {Controller#path_to}.
+  #
+  # = Reusing logic with hooks
+  #
+  # Methods can be defined and used as hooks for a route. For example:
+  #
+  #   Pakyow::App.router do
+  #     def foo
+  #     end
+  #
+  #     get :foo, "/foo", before: [:foo] do
+  #     end
+  #   end
+  #
+  # Before, after, and around hooks are supported in this way.
+  #
+  # = Extending routers
+  #
+  # Extensions can be defined and used to add shared routes to one or more
+  # routers. See {Routing::Extension}.
+  #
+  # = Other routing features
   #
+  # More advanced route features are available, including groups, namespaces,
+  # and templates. See {group}, {namespace}, and {template}.
+  #
+  # = Router subclasses
+  #
+  # It's possible to work with routers outside of Pakyow's DSL. For example:
+  #
+  #   class FooRouter < Pakyow::Router("/foo", after: [:bar])
+  #     def bar
+  #     end
+  #
+  #     default do
+  #       # available at GET /foo
+  #     end
+  #   end
+  #
+  #   Pakyow::App.router << FooRouter
+  #
+  # = Custom matchers
+  #
+  # Routers and routes can be defined with a matcher, which could be a +Regexp+ or
+  # any custom object that implements +match?+. For example:
+  #
+  #   class CustomMatcher
+  #     def match?(path)
+  #       path == "/custom"
+  #     end
+  #   end
+  #
+  #   Pakyow::App.router CustomMatcher.new do
+  #   end
+  #
+  # Custom matchers can also make data available in +params+ by implementing
```

Comment:
I have come up with a couple contrived uses, but nothing that specifically led me to add this feature. Supporting custom matchers is really just a side-effect of our implementation.